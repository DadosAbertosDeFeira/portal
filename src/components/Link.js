/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import NextLink from 'next/link';
import PropTypes from 'prop-types';

export default function Link({ children, href, ...rest }) {
  return (
    <NextLink href={href}>
      <a {...rest}>{children}</a>
    </NextLink>
  );
}

Link.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.string])
    .isRequired,
  href: PropTypes.string.isRequired,
};
