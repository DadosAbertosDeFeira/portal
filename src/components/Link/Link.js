/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import NextLink from 'next/link';
import PropTypes from 'prop-types';

const isExternal = (url) => {
  const isExternalRegex = new RegExp(/^(http|https):\/\//, 'i');
  return isExternalRegex.test(url);
};

export default function Link({ children, target, href, ...rest }) {
  const isExternalLink = isExternal(href);
  return isExternalLink ? (
    <a href={href} target={target} {...rest}>
      {children}
    </a>
  ) : (
    <NextLink href={href}>
      <a {...rest} target={target}>
        {children}
      </a>
    </NextLink>
  );
}
Link.defaultProps = {
  target: '_self',
};
Link.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.string,
    PropTypes.node,
  ]).isRequired,
  href: PropTypes.string.isRequired,
  target: PropTypes.string,
};
