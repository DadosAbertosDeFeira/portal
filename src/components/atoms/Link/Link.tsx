import type { ButtonVariants } from "atoms/Button";
import { buttonVariants } from "atoms/Button";
import NextLink from "next/link";
import type { ComponentPropsWithRef } from "react";
import { forwardRef, useEffect, useState } from "react";
import { twMerge } from "tailwind-merge";

export type LinkProps = ComponentPropsWithRef<typeof NextLink> & {
  variant?: ButtonVariants;
};

export const Link = forwardRef<HTMLAnchorElement, LinkProps>(function Link(
  { children, variant = "default", className, href, ...props },
  ref
) {
  const [isExternalURL, setExternalURL] = useState(false);

  useEffect(() => {
    try {
      const url = new URL(href as string);
      if (window) {
        setExternalURL(url.origin !== window.location.origin);
      }
    } catch (e) {
      setExternalURL(false);
    }
  }, [href]);

  return (
    <NextLink
      href={href}
      {...props}
      className={(twMerge(buttonVariants[variant]), className)}
      target={isExternalURL ? "_blank" : undefined}
      rel={isExternalURL ? "noreferrer" : undefined}
      ref={ref}
    >
      {children}
    </NextLink>
  );
});
