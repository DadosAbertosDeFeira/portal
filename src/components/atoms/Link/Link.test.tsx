import { render } from "@testing-library/react";

import { Link, type LinkProps } from "./Link";

describe("Link", () => {
  const makeSut = ({ href = "", ...props }: Partial<LinkProps>) => {
    return render(<Link href={href} {...props} />);
  };

  it("renders correctly", () => {
    const { asFragment } = makeSut({});

    expect(asFragment).toMatchSnapshot();
  });
});
