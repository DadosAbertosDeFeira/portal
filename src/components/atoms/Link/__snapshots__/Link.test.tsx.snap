// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders correctly 1`] = `[Function]`;
